### ISSUE TITLE
Show emacs init time

### ISSUE THREAD

**Issue by thorstengrothe** [emacs-dashboard/emacs-dashboard#57]
First of all, thanks for this package, I searched for a long time for something like this!

Is it possible to create a widget that shows the startup time of emacs?

Regards
Thorsten

**Comment by rakanalh** [comment]
Hi @thorstengrothe,

Yes, it's possible. The question is when. This is actually one of the things i had in mind. Can i encourage you to give it a go and open a PR? if you do, i'll let you know about my favorite Buckethead album (since you're a fan evident from your profile picture 😄 )

**Comment by thorstengrothe** [comment]
Hi,
I would help if I could 😄 but I'm a emacs beginner, but maybe I find something from someone else. The idea ist not new, spacemacs has it in it's dashboard AFAIK..

(And yes, I love Buckethead very much since a couple of years, especially the slow and mellow tracks and the ones with GNR, but that's another topic 😄)

**Comment by mclearc** [comment]
One thing to do is define a variable right at the start of your init file and then have dashboard load it (at least if you call dashboard at the end of your init file) as the banner message like so:

```emacs-lisp

(defvar my-init-el-start-time (current-time) "Time when init.el was started")

(setq dashboard-banner-logo-title (message " ★ Emacs initialized in %.2fs ★ " (float-time (time-subtract (current-time) my-init-el-start-time))))

```

**Comment by thorstengrothe** [comment]
@mclearc, thank you for the snippet it works pretty nice! But there is a difference between the loading time of your snippet and emacs-init-time,

Your snippet shows me: 1.11 sec.
Emacs init time: 1.5 sec. 

Which one is right? For convenience I show you my init.el here. My other config is in org-mode and I load it with org-babel, see here:

```
  ;; Init time start
(defvar my-init-el-start-time (current-time) "Time when init.el was started")

;; Faster startup
(setq gc-cons-threshold 100000000)

;; Package archives
(require 'package)
(setq package-archives '(("org"       . "http://orgmode.org/elpa/")
                         ("gnu"       . "http://elpa.gnu.org/packages/")
                         ("melpa"     . "https://melpa.org/packages/")
                         ;;("marmalade" . "http://marmalade-repo.org/packages/")
                         ))

(setq package-enable-at-startup nil)
(package-initialize)

;; Bootstrap use-package
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

(eval-when-compile
  (require 'use-package))
(require 'diminish)
(require 'bind-key)
;; (setq use-package-verbose t) ;; for erros

(let ((file-name-handler-alist nil)) ;; Startup time

  ;;; Load the config
  (org-babel-load-file (concat user-emacs-directory "thor_config.org"))
  ;;(load (concat user-emacs-directory "thor_config.elc"))

  ;; Speed up init time
  ;; https://www.reddit.com/r/emacs/comments/4j828f/til_setq_gcconsthreshold_100000000/
  (defun my-lower-gc-cons-threshold ()
    (setq gc-cons-threshold 800000)
    (remove-hook 'focus-out-hook #'my-lower-gc-cons-threshold))

  (add-hook 'after-init-hook
            (lambda ()
              (run-with-idle-timer
               60
               nil
               #'my-lower-gc-cons-threshold)
              (add-hook 'focus-out-hook #'my-lower-gc-cons-threshold)))
  
  (use-package dashboard
    :ensure t
    :diminish dashboard-mode
    :config
    (setq dashboard-banner-logo-title "Hi 😊 ")
    (setq dashboard-banner-logo-title
          (message " ★ Emacs initialized in %.2fs ★ "
                   (float-time (time-subtract (current-time) my-init-el-start-time))))
    (setq dashboard-startup-banner "~/.emacs.d/Bucketheaddeepwater.png")
    (setq dashboard-items '((agenda . 5)
                            (bookmarks . 10)
                            ))
    (dashboard-setup-startup-hook))
  )
```
Regards
Thorsten

**Comment by rakanalh** [comment]
Nice discussion... i think it might be better to use the `after-init-hook` as such:
```
(defvar startup-time (current-time))
(add-hook 'after-init-hook (lambda ()
                             (setq loading-time
                                   (float-time (time-subtract (current-time) startup-time)))
                             (message "%s" startup-time)))
```

**Comment by thorstengrothe** [comment]
@rakanalh with your snippet I got an error:

`custom-initialize-reset: Format specifier doesn’t match argument type`

this is slightly different and from jwiegleys config:

```
;; Init time start
(defconst emacs-start-time (current-time))

config stuff and org-babel to convert my init.el

(add-hook 'after-init-hook
            `(lambda ()
               (let ((elapsed
                      (float-time
                       (time-subtract (current-time) emacs-start-time))))
                 (message "Loading %s...done (%.2fs) [after-init]"
                          ,load-file-name elapsed))) t)
```
so far, so good now I got a message in my message buffer after init that shows me the correct startup time, but how do I get the output of the message in

`(setq dashboard-banner-logo-title ...` ?

**Comment by rememberYou** [comment]
Since the issue is still not closed, I would like to provide you with my solution in case it is still of interest to someone:
```
(use-package dashboard
  :if (< (length command-line-args) 2)
  :preface
  (defun my/dashboard-banner ()
    "Sets a dashboard banner including information on package initialization
     time and garbage collections."
    (setq dashboard-banner-logo-title
          (format "Emacs ready in %.2f seconds with %d garbage collections."
                  (float-time
                   (time-subtract after-init-time before-init-time)) gcs-done)))
  :init
  (add-hook 'after-init-hook 'dashboard-refresh-buffer)
  (add-hook 'dashboard-mode-hook 'my/dashboard-banner)
  :custom (dashboard-startup-banner 'logo)
  :config (dashboard-setup-startup-hook))
```

This is the result:
![2019-03-20_23:51:00](https://user-images.githubusercontent.com/6253527/54724351-16589a00-4b6b-11e9-8893-e93290f6c388.png)

**NOTE:** the `:if` clause prevents `dashboard` from blocking the opening of any file as an argument, when GNU Emacs is launched.

**Comment by jcs090218** [comment]
Seems like this feature is implemented after version `20190523.1449`?

**Comment by ema2159** [comment]
@JesusMtnez this should be closed
